packages/services/src/competition.service.ts

```diff
@@ -2290,6 +2290,54 @@ export class CompetitionService {
       );
     }
 
+    // Check blocklist FIRST - absolute exclusion that even VIPs cannot bypass
+    if (competition.blocklist?.includes(agentId)) {
+      throw new ApiError(
+        403,
+        "This agent is not permitted to join this competition",
+      );
+    }
+
+    // Check VIP status early - VIPs bypass soft requirements (stake, rank)
+    if (competition.vips && competition.vips.length > 0) {
+      if (competition.vips.includes(agentId)) {
+        try {
+          await this.competitionRepo.addAgentToCompetition(
+            competitionId,
+            agentId,
+          );
+        } catch (error) {
+          // Convert repository error to appropriate API error
+          if (
+            error instanceof Error &&
+            error.message.includes("maximum participant limit")
+          ) {
+            throw new ApiError(409, error.message);
+          }
+          // Handle one-agent-per-user error
+          if (
+            error instanceof Error &&
+            error.message.includes("already has an agent registered")
+          ) {
+            throw new ApiError(
+              409,
+              "You already have an agent registered in this competition. Each user can only register one agent per competition.",
+            );
+          }
+          throw error;
+        }
+        this.logger.debug(
+          `[CompetitionManager] VIP agent ${agentId} joined competition ${competitionId}, bypassing soft requirements`,
+        );
+        return;
+      }
+    }
+
+    // Validate participation rules (blocklist, allowlist, VIP, rank requirements)
```

Comment:
can remove blocklist from the comment

Reply:
Ah good call - removed here [5c96bcb](https://github.com/recallnet/js-recall/pull/1551/commits/5c96bcb76e8099d11c548ca08d732e6d93f56ff0)